File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    List<List<Integer>>ans;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        ans = new ArrayList<>();
        recur(candidates, 0, new ArrayList<>(), 0, target);
        return ans;
    }
    public void recur(int[] nums, int index, List<Integer>temp, int sum, int target){
        if(sum==target)
          ans.add(new ArrayList<>(temp));

        if (sum>target) return;
        
        if(index >= nums.length){
            return;
        }
        for(int i=index;i<nums.length;i++){
            temp.add(nums[i]);
            recur(nums, i, temp, sum+nums[i], target);
            temp.remove(temp.size()-1);
        }
    }
}

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>>ans;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int>temp;
        recur(candidates, 0,temp,target,0);
        return ans;
    }
    void recur(vector<int>& nums, int index, vector<int>&temp, int target, int sum){
        if(sum==target){
            ans.push_back(temp);
        }
        if(index >= nums.size() || sum>target){
            return;
        }
        for(int i=index;i<nums.size();i++){
            temp.push_back(nums[i]);
            recur(nums, i, temp,target, nums[i]+sum);
            temp.pop_back();
        }
    }
};